### PR TITLE
keepassxc: 2.6.6 -> 2.7.1

### LINKED ISSUE #165747: KeePassXC version 2.6.6 -> 2.7.0
###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: keepassxc

current version: 2.6.6
desired version: 2.7.0

###### Notify maintainers

maintainers: @jonafato @turion 

###### Note for maintainers

Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by Stunkymonkey**
###### Description of changes
new update [blog-post](https://keepassxc.org/blog/2022-03-21-2.7.0-released/) / [tag](https://github.com/keepassxreboot/keepassxc/releases/tag/2.7.0)
has not yet been tested on darwin, but before merging someone should do it.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jansol** [comment]
Word of warning, this release consistently fails to load some databases keepassxreboot/keepassxc#7590

**Comment by Kranzes** [comment]
Please test that it is actually not needed anymore.

**Comment by Stunkymonkey** [comment]
compiles without it. I am testing this now for a few days

**Comment by twz123** [comment]
I've also tested a bit on Linux:

* libsodium is also replaced by botan2.
* The yubikey libs are no longer required, as they have been replaced by libusb1 and pcsclite (at least on Linux).
* quazip has been [replaced](https://github.com/keepassxreboot/keepassxc/commit/12990e59ada1628a3a491afef045902ef254e38f) by minizip and WITH_XC_KEESHARE_SECURE is no longer a flag, but will be autodetected by the presence of minizip.
* WITH_XC_TOUCHID has been [removed](https://github.com/keepassxreboot/keepassxc/commit/4f0710350fb53c25db6824f96961d8d17a29347c)
* glibcLocales and libgpg-error are [not listed as dependencies](https://github.com/keepassxreboot/keepassxc/blob/2.7.0/INSTALL.md#build-dependencies) and for me it compiles and runs without them.

**Comment by twz123** [comment]
Not sure about the darwin patch, but if I got it right, it's about not invoking macdeloyqt. This has been [moved](https://github.com/keepassxreboot/keepassxc/commit/ab153a24ec2d5963192db63076bc0be1afa3ad95) to `src/post_install/CMakeLists.txt`, so maybe something like this is required, too?

```patch
diff --git a/src/post_install/CMakeLists.txt b/src/post_install/CMakeLists.txt
index 359c891f..a6a061a3 100644
--- a/src/post_install/CMakeLists.txt
+++ b/src/post_install/CMakeLists.txt
@@ -1,7 +1,7 @@
 # The install commands in this subdirectory will be executed after all the install commands in the
 # current scope are ran. It is required for correct functtioning of macdeployqt.
 
-if(APPLE AND WITH_APP_BUNDLE)
+if(FALSE)
     # Run macdeloyqt on the main app and any extra binaries and plugins as specified by the
     # _MACDEPLOYQT_EXTRA_BINARIES global property.
     # All install(TARGETS) calls should have already been called.
```

**Comment by mannp** [comment]
Result of `nixpkgs-review pr 166010` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>keepassxc</li>
  </ul>
</details>

**Comment by turion** [comment]
The integration seems to fail on x86_64-linux, at least that should be fixed. I can have a look into that later.

I don't understand why aarch64-darwin doesn't attempt to build.

**Comment by turion** [comment]
I took the liberty to fix the integration test on your branch. Everything else LGTM. We can have a look whether the test will run on darwin Hydra.

**Comment by turion** [comment]
@ofborg build keepassxc
@ofborg test keepassxc

**Comment by turion** [comment]
Doesn't build on darwin yet :sweat_smile: is the patch right that way?

**Comment by Stunkymonkey** [comment]
thanks for fixing the test. 

I do not own a darwin device to debug this easily. Maybe someone is interested to help.

**Comment by ellmau** [comment]
Note that keepassxc version 2.7.1 got released 2 days ago:

https://github.com/keepassxreboot/keepassxc/releases/tag/2.7.1

**Comment by Stunkymonkey** [comment]
@ofborg build keepassxc

File: pkgs/applications/misc/keepassx/community.nix [hunk 1]
```diff
@@ -95,26 +88,20 @@ stdenv.mkDerivation rec {
...
     qtx11extras
     readline
-    yubikey-personalization
     zlib
   ]
-  ++ optional withKeePassKeeShareSecure quazip
-  ++ optional stdenv.isDarwin qtmacextras
-  ++ optional (stdenv.isDarwin && withKeePassTouchID)
-    darwin.apple_sdk.frameworks.LocalAuthentication;
+  ++ optionals stdenv.isLinux [ libusb1 pcsclite ]
```
Comment (totoroot) [inline pkgs/applications/misc/keepassx/community.nix:103]:
```suggestion
  ++ optional stdenv.isLinux libusb1
```

File: pkgs/applications/misc/keepassx/community.nix [hunk 2]
```diff
@@ -95,26 +88,20 @@ stdenv.mkDerivation rec {
...
-    glibcLocales
+    botan2
     libXi
     libXtst
     libargon2
-    libgcrypt
-    libgpg-error
-    libsodium
-    libyubikey
+    minizip
```
Comment (totoroot) [inline pkgs/applications/misc/keepassx/community.nix:95]:
```suggestion
    minizip
    pcsclite
```

**Review by totoroot — CHANGES_REQUESTED** [review]
I tested the PR on my MBP and `pcsclite` seems to be needed on darwin as well.
However the build still fails due to issues I don't quite understand:
<details>
<summary>Build errors on darwin</summary>

```make
/tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:73:16:warning: comparison of integers of different signs: 'RETVAL' (aka 'unsigned int') and 'LONG'(aka 'int') [-Wsign-compare]
        if (rv == SCARD_E_SERVICE_STOPPED) {
            ~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:187:32: warning: comparison of integers of different signs: 'RETVAL' (aka 'unsigned int') and 'LONG' (aka 'int') [-Wsign-compare]
                    if (rv_act == SCARD_W_RESET_CARD) {
                        ~~~~~~ ^  ~~~~~~~~~~~~~~~~~~
/tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:495:17: error: return type 'int' must match previous return type 'unsigned int' when lambda expression has unspecified explicit return type
                return SCARD_E_FILE_NOT_FOUND;
                ^
/tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:473:16: error: no matching function for call to 'transactRetry'
        return transactRetry(handle.first, [&handle, &slot_cmd, &input, &output]() {
               ^~~~~~~~~~~~~
/tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:174:12: note: candidate function not viable: no known conversion from '(lambda at /tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:473:44)' to 'const std::function<RETVAL ()>' (aka 'const function<unsigned int ()>') for 2nd argument
    RETVAL transactRetry(SCARDHANDLE handle, const std::function<RETVAL()>& atomic_action)
           ^
/tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:745:16: warning: comparison of integers of different signs: 'unsigned int' and 'LONG' (aka 'int') [-Wsign-compare]
        if (rv == SCARD_W_CARD_NOT_AUTHENTICATED) {
            ~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-keepassxc-2.7.1.drv-1/source/src/keys/drivers/YubiKeyInterfacePCSC.cpp:748:23: warning: comparison of integers of different signs: 'unsigned int' and 'LONG' (aka 'int') [-Wsign-compare]
        } else if (rv == SCARD_E_FILE_NOT_FOUND) {
                   ~~ ^  ~~~~~~~~~~~~~~~~~~~~~~
4 warnings and 2 errors generated.
make[2]: *** [src/CMakeFiles/keepassx_core.dir/build.make:2687: src/CMakeFiles/keepassx_core.dir/keys/drivers/YubiKeyInterfacePCSC.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:650: src/CMakeFiles/keepassx_core.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
error: builder for '/nix/store/9y01hd1gbrpjia56f62w5ixhx44vbv71-keepassxc-2.7.1.drv' failed with exit code 2
```
</details>

In case you have an idea how to troubleshoot this, I can test any changes on my darwin machine.

**Comment by totoroot** [comment]
Build fails with some `zlib` issues that I couldn't successfully debug:
```make
ERROR: Feature 'system-zlib' was enabled, but the pre-condition 'libs.zlib' failed.
error: builder for '/nix/store/mgcim35aq56lhh0lx7v53zfx6fa812mq-qtbase-5.12.10.drv' failed with exit code 3;
       last 10 log lines:
       >
       > Note: When linking against OpenSSL, you can override the default
       > library names and directories with setting OPENSSL_PREFIX (or OPENSSL_INCDIR, OPENSSL_LIBDIR),
       > and OPENSSL_LIBS, OPENSSL_LIBS_DEBUG, OPENSSL_LIBS_RELEASE.
       > For example:
       >     ./configure -openssl-linked OPENSSL_PREFIX='/opt/ssl' OPENSSL_LIBS='-lssl -lcrypto'
       >
       > WARNING: On OS X, AAT is supported only with -qt-harfbuzz.
       >
       > ERROR: Feature 'system-zlib' was enabled, but the pre-condition 'libs.zlib' failed.
       For full logs, run 'nix log /nix/store/mgcim35aq56lhh0lx7v53zfx6fa812mq-qtbase-5.12.10.drv'.
error: 1 dependencies of derivation '/nix/store/vs06h1y80sqqw29mba04xgws775f7jn1-keepassxc-2.7.1.drv' failed to build
```

Perhaps someone more knowledgeable than me has an idea...

**Comment by turion** [comment]
I think your last error is https://github.com/NixOS/nixpkgs/issues/167708, which should be fixed. Can you rebase and try again?

**Comment by totoroot** [comment]
Thanks! I would also like to see this merged before 22.05 feature freeze. Today I got around to testing on darwin again.

I finally managed to build it on darwin by doing the following:

- Enabling YubiKey for linux only as on darwin the issues that I described in an earlier reply persist
I'm not sure if it could be merged like this, since removing YubiKey support for darwin users might not be welcomed.
- Adding support for Touch ID again as it caused the build to fail
While the flag `WITH_XC_TOUCHID` has been removed with the new release of KeepassXC as described above, we still have to include `darwin.apple_sdk.frameworks.LocalAuthentication` on darwin.

- Disabling `testautotype` as it failed with the following:

<details>
<summary>Click to expand failing tests</summary>

```make
25/37 Test #18: testautotype .....................Subprocess aborted***Exception:   3.59 sec
********* Start testing of TestAutoType *********
Config: Using QtTest library 5.14.2, Qt 5.14.2 (x86_64-little_endian-lp64 shared(dynamic) release build; by Clang 11.1.0 )
PASS   : TestAutoType::initTestCase()
PASS   : TestAutoType::testInternal()
QFATAL : TestAutoType::testSingleAutoType() QWidget: Cannot create a QWidget without QApplication
FAIL!  : TestAutoType::testSingleAutoType() Received a fatal error.
   Loc: [Unknown file(0)]
Totals: 2 passed, 1 failed, 0 skipped, 0 blacklisted, 1733ms
********* Finished testing of TestAutoType *********

      Start 37: testbrowser
26/37 Test #28: testbase32 .......................   Passed    1.59 sec
27/37 Test #29: testcsvparser ....................   Passed    1.53 sec
28/37 Test #30: testrandomgenerator ..............   Passed    1.32 sec
29/37 Test #31: testentrysearcher ................   Passed    1.46 sec
30/37 Test #32: testcsvexporter ..................   Passed    1.54 sec
31/37 Test #33: testsharing ......................   Passed    1.58 sec
32/37 Test #35: testtools ........................   Passed    1.60 sec
33/37 Test #36: testconfig .......................   Passed    1.62 sec
34/37 Test #37: testbrowser ......................   Passed    1.76 sec
35/37 Test #34: testdatabase .....................   Passed    2.16 sec
36/37 Test  #4: testkdbx4 ........................   Passed    8.13 sec
37/37 Test #12: testmodified .....................   Passed   10.82 sec

97% tests passed, 1 tests failed out of 37

Total Test time (real) =  10.85 sec

The following tests FAILED:
         18 - testautotype (Subprocess aborted)
Errors while running CTest
make: *** [Makefile:91: test] Error 8
error: builder for '/nix/store/b8kh02ym280xqpyz13spbl1mhvrp987l-keepassxc-2.7.1.drv' failed with exit code 2;
       last 10 log lines:
       > 37/37 Test #12: testmodified .....................   Passed   10.82 sec
       >
       > 97% tests passed, 1 tests failed out of 37
       >
       > Total Test time (real) =  10.85 sec
       >
       > The following tests FAILED:
       >       18 - testautotype (Subprocess aborted)
       > Errors while running CTest
       > make: *** [Makefile:91: test] Error 8
       For full logs, run 'nix log /nix/store/b8kh02ym280xqpyz13spbl1mhvrp987l-keepassxc-2.7.1.drv'.
```
</details>

Also I believe that `DWITH_XC_AUTOTYPE` can be removed, since it defaults to `ON` anyways, right?

You can take a look at the changes on [my branch](https://github.com/totoroot/nixpkgs/tree/update/keepassxc-2.7.1).

**Comment by totoroot** [comment]
Autotype works just fine so I'm not sure why this particular test is failing.
According to the [installation docs](https://github.com/keepassxreboot/keepassxc/blob/develop/INSTALL.md) apart from `qtx11extras`, `libxi`, and `libxtst` (for auto-type on X11) no additional libraries are needed on macOS

**Comment by turion** [comment]
> Would like to get this merged into 22.05.

That would be great :)  but it has to work on all platforms again for that.

> Enabling YubiKey for linux only as on darwin the issues that I described in an earlier reply persist
>
> I'm not sure if it could be merged like this, since removing YubiKey support for darwin users might not be welcomed.

That sounds like a big feature that shouldn't be removed.

> According to the [installation docs](https://github.com/keepassxreboot/keepassxc/blob/develop/INSTALL.md) apart from `qtx11extras`, `libxi`, and `libxtst` (for auto-type on X11) no additional libraries are needed on macOS

I don't understand, according to https://github.com/keepassxreboot/keepassxc/wiki/Set-up-Build-Environment-on-macOS you need more dependencies:
```
brew install qt5 argon2 botan qrencode asciidoctor readline minizip
```

**Comment by droidmonkey** [comment]
For your pcsc errors you might need to adjust this to account for nixos: https://github.com/keepassxreboot/keepassxc/blob/f15ee904298d62ef5bda0cd8eda9ace45d7288ff/src/keys/drivers/YubiKeyInterfacePCSC.cpp#L34

**Comment by turion** [comment]
I'll have a look into the failing integration test.

**Comment by turion** [comment]
The integration test works on my machine :thinking: 

Tested on fb0a5839170561f4d752ed96fb4856f2d3a10c5d with 

```
$ nix-build . -A nixosTests.mytest.driverInteractive
$ ./result/bin/nixos-test-driver --interactive
```

**Comment by droidmonkey** [comment]
The error likely is happening here: https://github.com/keepassxreboot/keepassxc/blob/f15ee904298d62ef5bda0cd8eda9ace45d7288ff/src/autotype/AutoType.cpp#L338

We missed a guard to make sure a main window was present before trying to show an error message box. Although... why there is an error displaying for this test is another matter entirely.

**Comment by turion** [comment]
I just realized that apparently keepassxc doesn't build on a mac since a few months: https://hydra.nixos.org/build/175226618

Is this the same (or similar) build failure that we have here?

**Comment by mannp** [comment]
> I just realized that apparently keepassxc doesn't build on a mac since a few months: https://hydra.nixos.org/build/175226618
> 
> Is this the same (or similar) build failure that we have here?

Is it time to release this nice update with;

```
meta = {
  # ref to issue/explanation
  # `true` is for everything
  broken = stdenv.isDarwin; 
};
```

**Comment by Stunkymonkey** [comment]
I am highly in favor of merging this.
The current status does not compile on darwin and with this PR it continues to work for `linux`.

Since I do not own a `darwin` device it is quite hard/impossible work to figure out why it does not.

So my idea is to merge this and someone with a darwin device proposes another PR to fix the compilation.

**Comment by turion** [comment]
Agree. Are there any open discussion points that need addressing? (I can't seem to find any more.) Let me quickly check why the integration test fails.

**Comment by turion** [comment]
> Let me quickly check why the integration test fails.

OCR has troubles with the inverted colors. Let me simplify the test.

**Comment by turion** [comment]
I took the liberty to push the changes on your branch. If it goes through, I'll merge.

**Pull request merged**